opm/core/grid/RepairZCORN.hpp

```diff
@@ -0,0 +1,728 @@
+                        const CartDims&         cartDims)
+                : nx_(cartDims[0])
+                , ny_(cartDims[1])
+                , nz_(cartDims[2])
+            {
+                const auto nglob = nx_ * ny_ * nz_;
+
+                if (actnum.empty()) {
+                    this->is_active_.resize(nglob, true);
+                    this->acells_   .resize(nglob, 0);
+
+                    std::iota(std::begin(this->acells_),
+                              std::end  (this->acells_), 0);
+                }
+                else if (actnum.size() == nglob) {
+                    this->is_active_.resize(nglob, false);
+                    this->acells_   .reserve(nglob);
+
+                    for (auto i = 0*nglob; i < nglob; ++i) {
+                        if (actnum[i] != 0) {
+                            this->acells_.push_back(i);
+                            this->is_active_[i] = true;
+                        }
+                    }
+                }
+                else {
+                    throw std::invalid_argument {
+                        "ACTNUM vector does not match global size"
+                    };
+                }
+            }
+
+            /// Representation of Cartesian cell position.
+            using IndexTuple = std::array<std::size_t, 3>;
+
+            /// Retrieve cell position (\code (I, J, K) \endcode tuple) in
+            /// global grid.
+            ///
+            /// \param[in] globCell Global cell ID of particular cell.
+            ///
+            /// \return Cartesian index position of \p globCell.
+            IndexTuple getCellIJK(const std::size_t globCell) const
+            {
+                auto c = globCell;
+
+                auto i = c % nx_;  c /= nx_;
+                auto j = c % ny_;  c /= ny_;
+                auto k = c % nz_;
+
+                return {{ i, j, k }};
+            }
+
+            /// Retrieve global IDs of model's explicitly active cells.
+            const std::vector<std::size_t>& activeGlobal() const
+            {
+                return this->acells_;
+            }
+
+            /// Retrieve total number of uncompressed cells in model.
+            std::size_t numGlobalCells() const
+            {
+                return this->nx_ * this->ny_ * this->nz_;
+            }
+
+            /// Retrieve global ID of cell immediately below particular
+            /// cell.
+            ///
+            /// Assumes that lateral cell positions are valid.
+            ///
+            /// \param[in] ijk Cartesian index position of single cell.
+            ///
+            /// \return Global ID of cell immediately below \p ijk.  If no
+            ///    such cell exists, e.g., because the input cell is in the
+            ///    bottom layer of the model or all cells below it are
+            ///    explicitly deactivated, then function neighbourBelow()
+            ///    returns \code static_cast<std::size_t>(-1) \endcode.
```

Comment:
Can the check `i < numGlobalCells()` become false for any *other* reason than the above? If not I think it should be signed.